{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process new SAMOSA samples\n",
    "\n",
    "This code takes in CSV files from the Ramani Lab google drive pacbio run set ups and outputs scripts to do the read processing on Wynton, as well as a fasta file of the barcodes and a sample reference CSV.\n",
    "\n",
    "Instructions:\n",
    "1. Find the sheet for the run you want to process in the google drive, under Ramani Lab > instruments > Pacbio > runs > Run Set Ups\n",
    "2. Download the first and last tabs (*Samples* and *Demux FASTA files*) as CSV files\n",
    "3. Transfer these CSV files files to the UCSF server where you are running this notebook.\n",
    "4. Run this notebook. Adjusted the hardcoded name checks as needed to include your cell and use the correct reference\n",
    "5. If not running on wynton, transfer the output (scripts, barcodes.fasta, and samplereference.csv) to Wynton\n",
    "6. qsub the runLima.sh script for your cell\n",
    "7. If it's an amplicon reference, you will then need to qsub the runExtractAll.sh script after that script and daughter scripts are finished running\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import socket\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "if 'biochem1' in socket.gethostname():\n",
    "    os.chdir('/avicenna/cmcnally/pipeline')\n",
    "    # pipeline is the folder I used to contain the input and output from this notebook. You could make your\n",
    "    # own pipeline folder and use it in a similar way\n",
    "    \n",
    "    dataPBase = '/avicenna/vramani/analyses/pacbio/'\n",
    "    rawPBase = '/avicenna/vramani/data/pacbio/'\n",
    "    figPBase = '/avicenna/cmcnally/pbanalysis/'\n",
    "if 'titan' in socket.gethostname():\n",
    "    dataPBase = '/data/users/goodarzilab/colin/results/pacbio/'\n",
    "if 'wynton' in socket.gethostname():\n",
    "    dataPBase = '/wynton/group/goodarzilab/ramanilab/results/pacbio/'\n",
    "    rawPBase = '/wynton/group/goodarzilab/ramanilab/data/pacbio/'\n",
    "if 'rumi' in socket.gethostname():\n",
    "    raise Exception('no pacbio results folder on rumi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "csvfs = os.listdir('gdoc_sheets') # looks for files in the gdoc_sheets directory within pipeline\n",
    "sampfs = [cv for cv in csvfs if 'Samples.csv' in cv] # Takes all the Samples csv files\n",
    "\n",
    "# Uses the assumption that the order of the cells in the csv is the same as the order they were sequenced on\n",
    "# the machine. This is usually true but be careful of possible exceptions to this assumption\n",
    "cellIfolder = {1:'1_B03', 2:'2_B05', 3:'3_B07', 4:'4_B09'}\n",
    "\n",
    "# Goes through all of the Samples csv files and generates output scripts, fasta, and samplereference for them\n",
    "# This could be modified to only generate the output for a single new input\n",
    "for i in range(len(sampfs)):\n",
    "    cellNames = []\n",
    "    cellInd = {}\n",
    "    cellI = 1\n",
    "    with open('gdoc_sheets/{0}'.format(sampfs[i])) as csvfile:\n",
    "        cvreader = csv.reader(csvfile, delimiter=',')\n",
    "        inSamples = False\n",
    "        for row in cvreader:\n",
    "            \n",
    "            # The script looks for a header field called Directory Name and takes the entry as the folder name\n",
    "            # of the sequencing output. This will look something like r64182_20211123_032844\n",
    "            # This field may not be present in the google doc. If it isn't, you need to look up the folder name\n",
    "            # and add it to the google doc before you download the Samples sheet\n",
    "            if row[0] == 'Directory Name' or row[0] == 'Directory Name:':\n",
    "                runName = row[1]\n",
    "            if inSamples:\n",
    "                if len(row[1]) > 3:\n",
    "                    # Keep track of how many cells have been seen\n",
    "                    cellInd[row[1]] = cellI\n",
    "                    cellI += 1\n",
    "                    \n",
    "                if np.array([a in row[1] for a in {'SAM', '_NA_', 'BrdU', 'dUTP'}]).any():\n",
    "                    # If any of these substrings are in the cell name, assume it is SAMOSA and make output for it\n",
    "                    cellNames.append(row[1])\n",
    "            if row[1] == 'Pool / Run Name' or row[1] == 'Pool / Cell Name':\n",
    "                inSamples = True\n",
    "\n",
    "    assert runName != '', \"No directory for the files\"\n",
    "    \n",
    "    if cellNames[0] == '210421_NA_SAMv2_mESCs': #fix for one cell that didn't have the directory name field\n",
    "        runName = 'r64182_20210421_233732'\n",
    "        \n",
    "    cellSamples = {}\n",
    "    # Now open the other sheet and from it take the sample names and barcodes\n",
    "    with open('gdoc_sheets/{0}'.format(sampfs[i].replace('Samples', 'Demux FASTA files'))) as csvfile:\n",
    "        cvreader = csv.reader(csvfile, delimiter=',')\n",
    "        for row in cvreader:\n",
    "            if row[0] in cellNames:\n",
    "                cellSamples[row[0]] = row[1]\n",
    "                \n",
    "    for cell in cellNames:\n",
    "        sampleNames = []\n",
    "        barcodes = []\n",
    "\n",
    "        # get sample names and barcodes, clean up sample names\n",
    "        barcElements = cellSamples[cell].split('\\n')\n",
    "        for b in barcElements:\n",
    "            if '>' in b:\n",
    "                sampn = b\n",
    "                sampn = sampn.replace('>','')\n",
    "                # Replace problematic symbols in the sample names. These names will be used as file names and\n",
    "                # some cause issues with the Pacbio software or unix in general\n",
    "                sampn = sampn.replace(' ', '_')\n",
    "                sampn = sampn.replace('+m', 'plusM')\n",
    "                sampn = sampn.replace('+M', 'plusM')\n",
    "                sampn = sampn.replace('-m', 'minusM')\n",
    "                sampn = sampn.replace('-M', 'minusM')\n",
    "                sampn = sampn.replace('+ATP', 'plusATP')\n",
    "                sampn = sampn.replace('+ADP', 'plusADP')\n",
    "                sampn = sampn.replace(':', 'to')\n",
    "                sampn = sampn.replace('.', '_') #dislike this one, better to avoid in first place\n",
    "                if sampn[-1] in ['1','2'] and 'rep' not in sampn:\n",
    "                    sampn = sampn[:-1] + '_rep' + sampn[-1]\n",
    "                sampn = sampn.rstrip(' _')\n",
    "                sampleNames.append(sampn)\n",
    "            else:\n",
    "                if b is not '':\n",
    "                    barcodes.append(b)\n",
    "        if len(sampleNames) != len(barcodes):\n",
    "            raise Exception('sample names and barcodes not matching up')\n",
    "        \n",
    "        # write out barcodes to a fasta\n",
    "        with open('/avicenna/cmcnally/pipeline/barcodes/{0}.barcodes.fasta'.format(cell), 'w') as fout:\n",
    "            for i in range(len(sampleNames)):\n",
    "                fout.write('>' + sampleNames[i] + '\\n')\n",
    "                fout.write(barcodes[i] + '\\n')\n",
    "\n",
    "\n",
    "        if not os.path.exists('/avicenna/cmcnally/pipeline/scripts/' + cell):\n",
    "            os.makedirs('/avicenna/cmcnally/pipeline/scripts/' + cell)\n",
    "\n",
    "        # write out the runLima.sh script. This will make the output folder (OUTDIR), run lima, and then start running\n",
    "        # the individual scripts to process each sample separately after lima completes\n",
    "        with open('/avicenna/cmcnally/pipeline/scripts/{0}/runLima.sh'.format(cell), 'w') as fout:\n",
    "            fout.write('#!/bin/env bash\\n#\\n#$ -pe smp 24\\n#$ -l h_rt=72:00:00\\n\\n\\n')\n",
    "            fout.write('INFILE=/wynton/group/goodarzilab/ramanilab/data/pacbio/{0}/{1}/m64182_*.subreads.bam\\n'.format(runName, cellIfolder[cellInd[cell]]))\n",
    "            fout.write('OUTDIR=/wynton/group/goodarzilab/ramanilab/results/pacbio/{0}\\n\\n'.format(cell))\n",
    "            fout.write('mkdir -p ${OUTDIR}/{aligned,ccs,processed}\\n')\n",
    "            fout.write('cd ${OUTDIR}\\n\\n')\n",
    "\n",
    "            fout.write('cp /wynton/home/goodarzi/cpmcnally/pipeline/barcodes/{0}.barcodes.fasta ${{OUTDIR}}\\n\\n'.format(cell))\n",
    "            fout.write('/wynton/home/goodarzi/cpmcnally/bin/smrtlink/smrtcmds/bin/lima --same --split-bam-named ')\n",
    "            fout.write('-j ${NSLOTS:-1} ${INFILE} ${OUTDIR}/')\n",
    "            fout.write('{0}.barcodes.fasta ${{OUTDIR}}/{0}.split.bam\\n\\n'.format(cell))\n",
    "            for samp in sampleNames:\n",
    "                fout.write('qsub /wynton/home/goodarzi/cpmcnally/pipeline/scripts/{0}/process_{1}.sh\\n'.format(cell, samp))\n",
    "\n",
    "\n",
    "        # prepare for sample reference\n",
    "        targethost = 'wynton'\n",
    "        if targethost == 'wynton':\n",
    "            pathbase = '/wynton/group/goodarzilab/ramanilab/results/pacbio'\n",
    "\n",
    "        refd = {'cell':[], 'sampleName':[], 'unalignedSubreadsFile':[], 'ccsFile':[], 'alignedSubreadsFile':[],\n",
    "                'alignedCcsFile':[], 'reference':[], 'processed':[]}\n",
    "                \n",
    "        # write extractIPD script for the cell. Add a line for each sample within the below loop\n",
    "        extractFile = open('/avicenna/cmcnally/pipeline/scripts/{0}/runExtractAll.sh'.format(cell), 'w')\n",
    "        extractFile.write('#!/usr/bin/env bash\\n\\n')\n",
    "        \n",
    "        # Write script for each sample, to make ccs and align subreads\n",
    "        for isamp, samp in enumerate(sampleNames):\n",
    "            genomicRef = False\n",
    "            ampliconRef = False\n",
    "            \n",
    "            # Figure out whether this sample is genomic or amplicon, and what the correct reference is\n",
    "            # Obviously this needs to be manually adjusted to make sure it correctly identifies your cells\n",
    "            if 'mESC' in samp or 'E14' in samp or '211122_NA_Snf2h' in cell:\n",
    "                genomicRef = True\n",
    "                refmmi = '/wynton/group/goodarzilab/ramanilab/genomes/mm10/mm10.mmi'\n",
    "            if 'K562' in samp:\n",
    "                genomicRef = True\n",
    "                refmmi = '/wynton/group/goodarzilab/ramanilab/genomes/hg38/hg38.mmi'\n",
    "\n",
    "            if '_observed' in samp or 'array' in samp or 'CTCF' in samp:\n",
    "                ampliconRef = True\n",
    "                if samp[0:5].lower() == 'depen':\n",
    "                    reffasta = 'snf2h_dependent_site_observed.fasta'\n",
    "                elif samp[0:8] == 'CTCF_Dep':\n",
    "                    reffasta = 'snf2h_dependent_site_observed.fasta'\n",
    "                elif samp[0:5].lower() == 'indep':\n",
    "                    reffasta = 'snf2h_independent_site_observed.fasta'\n",
    "                elif samp[0:8] == 'CTCF_Ind':\n",
    "                    reffasta = 'snf2h_independent_site_observed.fasta'\n",
    "                else:\n",
    "                    print('Ref unknown')\n",
    "                    print(samp)\n",
    "                    \n",
    "            if 'MO_dUTP_PCR' in cell:\n",
    "                ampliconRef = True\n",
    "                reffasta = 'snf2h_independent_site_observed.fasta'\n",
    "                \n",
    "            if not genomicRef and not ampliconRef:\n",
    "                print('Reference unknown: %s' % (samp))\n",
    "                # don't write output if the cell doesn't get identified properly\n",
    "                continue\n",
    "\n",
    "            with open('/avicenna/cmcnally/pipeline/scripts/{0}/process_{1}.sh'.format(cell, samp), 'w') as fout:\n",
    "                if len(sampleNames) >= 4:\n",
    "                    fout.write('#!/bin/env bash\\n#\\n#$ -pe smp 8\\n#$ -l h_rt=48:00:00\\n')\n",
    "                else: # if small number of samples, request more computational resources to process each one\n",
    "                    fout.write('#!/bin/env bash\\n#\\n#$ -pe smp 16\\n#$ -l h_rt=72:00:00\\n')\n",
    "                fout.write('#$ -l hostname=\"!qb3-ad*\"\\n\\n\\n') # request to not use nodes that don't support SSE4.1, and thus pbmm2 won't work on\n",
    "                fout.write('cd /wynton/group/goodarzilab/ramanilab/results/pacbio/{0}\\n\\n'.format(cell))\n",
    "\n",
    "                fout.write('/wynton/home/goodarzi/cpmcnally/bin/smrtlink/smrtcmds/bin/ccs -j ${NSLOTS:-1} ')\n",
    "                fout.write('{0}.split.{1}--{1}.bam ccs/{0}.split.{1}.ccs.bam\\n\\n'.format(cell,samp))\n",
    "                fout.write('/wynton/home/goodarzi/cpmcnally/bin/smrtlink/smrtcmds/bin/pbmm2 align ')\n",
    "                if genomicRef:\n",
    "                    fout.write('{2} ccs/{0}.split.{1}.ccs.bam aligned/{0}.split.{1}.ccs.aligned.sorted.bam '.format(cell,samp,refmmi))\n",
    "                    fout.write('--sort --preset CCS -j ${NSLOTS:-1}\\n\\n')\n",
    "                    fout.write('/wynton/home/goodarzi/cpmcnally/bin/smrtlink/smrtcmds/bin/pbindex aligned/{0}.split.{1}.ccs.aligned.sorted.bam\\n'.format(cell,samp))\n",
    "                    fout.write('\\nqsub /wynton/home/goodarzi/cpmcnally/pipeline/scripts/runExtractGenomicLinear.sh {0} {1}\\n'.format(cell, isamp))\n",
    "                    \n",
    "                    extractFile.write('qsub /wynton/home/goodarzi/cpmcnally/pipeline/scripts/runExtractGenomicLinear.sh {0} {1}\\n'.format(cell, isamp))\n",
    "                elif ampliconRef:\n",
    "                    fout.write('%s/pbrun10_CTCFpool_2/%s ' % (pathbase,reffasta))\n",
    "                    fout.write('{0}.split.{1}--{1}.bam aligned/{0}.split.{1}.subreads.aligned.bam '.format(cell,samp))\n",
    "                    fout.write('-j ${NSLOTS:-1}\\n\\n')\n",
    "                    fout.write('/wynton/home/goodarzi/cpmcnally/bin/smrtlink/smrtcmds/bin/pbindex aligned/{0}.split.{1}.subreads.aligned.bam\\n'.format(cell,samp))\n",
    "                    \n",
    "                    extractFile.write('qsub /wynton/home/goodarzi/cpmcnally/pipeline/scripts/runExtractAmplicon.sh {0} {1}\\n'.format(cell, isamp))\n",
    "            # also write the sample reference entry for this sample\n",
    "            refd['cell'].append(cell)\n",
    "            refd['sampleName'].append(samp)\n",
    "            refd['unalignedSubreadsFile'].append('{0}/{1}/{1}.split.{2}--{2}.bam'.format(pathbase, cell, samp))\n",
    "            refd['ccsFile'].append('%s/%s/ccs/%s.split.%s.ccs.bam' % (pathbase,cell, cell, samp))\n",
    "            if genomicRef:\n",
    "                refd['alignedSubreadsFile'].append('')\n",
    "                refd['alignedCcsFile'].append('{0}/{1}/aligned/{1}.split.{2}.ccs.aligned.sorted.bam'.format(pathbase, cell, samp))\n",
    "                refd['reference'].append(refmmi)\n",
    "            elif ampliconRef:\n",
    "                refd['alignedSubreadsFile'].append('%s/%s/aligned/%s.split.%s.subreads.aligned.bam' %\n",
    "                                                   (pathbase, cell, cell, samp))\n",
    "                refd['alignedCcsFile'].append('')\n",
    "                refd['reference'].append('%s/pbrun10_CTCFpool_2/%s' % (pathbase,reffasta))\n",
    "            refd['processed'].append('%s/%s/processed' % (pathbase,cell))\n",
    "            \n",
    "        extractFile.close()\n",
    "        # compile and write out sample reference csv\n",
    "        sref = pd.DataFrame(refd)\n",
    "        sref.to_csv('/avicenna/cmcnally/pipeline/sampleReferences/%s.sampleReference.wynton.csv' % (cell), index_label='index')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell writes out shell wrappers for the extractIPD script. This only needs to be run once as these\n",
    "# are generic shell scripts. The first one\n",
    "\n",
    "# write generic extractIPD script for genomic input\n",
    "\n",
    "with open('/avicenna/cmcnally/pipeline/scripts/runExtractGenomicLinear.sh', 'w') as fout:\n",
    "    fout.write('#!/usr/bin/env bash\\n#\\n#$ -pe smp 10\\n#$ -l mem_free=1G\\n#$ -l h_rt=96:00:00\\n#$ -l hostname=\"!qb3-ad*\"\\n\\n\\n')\n",
    "    fout.write('%s\\n%s\\n%s\\n' % ('# >>> conda initialize >>>',\n",
    "                                 '# !! Contents within this block are managed by ''conda init'' !!',\n",
    "                                 '__conda_setup=\"$(''/wynton/home/goodarzi/cpmcnally/bin/miniconda3/bin/conda'' ''shell.bash'' ''hook'' 2> /dev/null)\"'))\n",
    "    fout.write('%s\\n%s\\n%s\\n%s\\n%s\\n' % ('if [ $? -eq 0 ]; then',\n",
    "                                         '\\teval \"$__conda_setup\"',\n",
    "                                         'else',\n",
    "                                         '\\tif [ -f \"/wynton/home/goodarzi/cpmcnally/bin/miniconda3/etc/profile.d/conda.sh\" ]; then',\n",
    "                                         '\\t\\t. \"/wynton/home/goodarzi/cpmcnally/bin/miniconda3/etc/profile.d/conda.sh\"'))\n",
    "    fout.write('%s\\n%s\\n%s\\n%s\\n%s\\n%s\\n\\n' % ('\\telse',\n",
    "                                     '\\t\\texport PATH=\"/wynton/home/goodarzi/cpmcnally/bin/miniconda3/bin:$PATH\"',\n",
    "                                     '\\tfi',\n",
    "                                     'fi',\n",
    "                                     'unset __conda_setup',\n",
    "                                     '# <<< conda initialize <<<'))\n",
    "    fout.write('cd /wynton/group/goodarzilab/ramanilab/results/pacbio/$1\\n')\n",
    "    fout.write('conda activate procSAMOSA\\n\\n')\n",
    "    fout.write('python /wynton/home/goodarzi/cpmcnally/code/scripts/extractIPDlinear.py -j 8 -o /wynton/group/goodarzilab/ramanilab/results/pacbio/')\n",
    "    fout.write('$1 /wynton/home/goodarzi/cpmcnally/pipeline/sampleReferences/$1.sampleReference.wynton.csv $2\\n')\n",
    "    \n",
    "# write generic extractIPD script for amplicon input\n",
    "\n",
    "with open('/avicenna/cmcnally/pipeline/scripts/runExtractAmplicon.sh', 'w') as fout:\n",
    "    fout.write('#!/usr/bin/env bash\\n#\\n#$ -pe smp 1\\n#$ -l mem_free=12G\\n#$ -l h_rt=48:00:00\\n\\n\\n')\n",
    "    fout.write('%s\\n%s\\n%s\\n' % ('# >>> conda initialize >>>',\n",
    "                                 '# !! Contents within this block are managed by ''conda init'' !!',\n",
    "                                 '__conda_setup=\"$(''/wynton/home/goodarzi/cpmcnally/bin/miniconda3/bin/conda'' ''shell.bash'' ''hook'' 2> /dev/null)\"'))\n",
    "    fout.write('%s\\n%s\\n%s\\n%s\\n%s\\n' % ('if [ $? -eq 0 ]; then',\n",
    "                                         '\\teval \"$__conda_setup\"',\n",
    "                                         'else',\n",
    "                                         '\\tif [ -f \"/wynton/home/goodarzi/cpmcnally/bin/miniconda3/etc/profile.d/conda.sh\" ]; then',\n",
    "                                         '\\t\\t. \"/wynton/home/goodarzi/cpmcnally/bin/miniconda3/etc/profile.d/conda.sh\"'))\n",
    "    fout.write('%s\\n%s\\n%s\\n%s\\n%s\\n%s\\n\\n' % ('\\telse',\n",
    "                                     '\\t\\texport PATH=\"/wynton/home/goodarzi/cpmcnally/bin/miniconda3/bin:$PATH\"',\n",
    "                                     '\\tfi',\n",
    "                                     'fi',\n",
    "                                     'unset __conda_setup',\n",
    "                                     '# <<< conda initialize <<<'))\n",
    "    fout.write('cd /wynton/group/goodarzilab/ramanilab/results/pacbio/$1\\n')\n",
    "    fout.write('conda activate procSAMOSA\\n\\n')\n",
    "    fout.write('python /wynton/home/goodarzi/cpmcnally/code/scripts/extractIPDfull3.py -o /wynton/group/goodarzilab/ramanilab/results/pacbio/$1 ')\n",
    "    fout.write('/wynton/home/goodarzi/cpmcnally/pipeline/sampleReferences/$1.sampleReference.wynton.csv $2\\n')\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py4",
   "language": "python",
   "name": "py4"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
